### PR TITLE
Datasets + DataLoaders

### PR DESCRIPTION
**Pull request by hitorilabs**
Trying to build up some intuitions about traditional machine learning and deep learning techniques from scratch - with a focus around expressing it neatly in terms of code + visualizations.

Preferably not using any notebooks, just runnable `make` commands.

---

One specific path I want to go is a depth-first exploration from stable diffusion down to the foundational concepts that it is built on.

### CODE REVIEW

**Comment by hitorilabs** [comment]
very inaccurate, but somewhat useful to know in practice - closest *reading* I could get to max BW in raw CUDA is ~24GiB, which is not nearly theoretical maximum of PCIe 4.0 x16

```
python3 naive.py
completed in 0.06398421s | throughput=0.14022472499855704GiB/s
completed in 0.00047826s | throughput=18.759905774907672GiB/s
completed in 0.00046302s | throughput=19.377580837624777GiB/s
completed in 0.00045833s | throughput=19.575824939497323GiB/s
completed in 0.00046212s | throughput=19.41527600456674GiB/s

python3 preallocate_on_gpu.py
completed in 0.00051289s | throughput=17.4933911495675GiB/s
completed in 0.00046213s | throughput=19.414772242060117GiB/s
completed in 0.00045613s | throughput=19.6701131964522GiB/s
completed in 0.00045330s | throughput=19.79322017749102GiB/s
completed in 0.00045695s | throughput=19.635071544645644GiB/s

python3 load_from_disk.py
completed in 0.00356302s | throughput=2.518136408785562GiB/s
completed in 0.00292387s | throughput=3.06859541002718GiB/s
completed in 0.00296910s | throughput=3.021843575715799GiB/s
completed in 0.00118152s | throughput=7.5937694905407875GiB/s
completed in 0.00102452s | throughput=8.757444391834103GiB/s
```
For disk load, it's approximately limited by it's own data rate `7300MB/s (7.1GiB/s)`

RAM is something like:
```
bytes_per_transfer = 8  # 64-bit (or 8-byte) data bus
clock_speed_mhz = 5600
channels = 2

# Calculate bandwidth in bytes per second
bandwidth_bytes_per_second = clock_speed_mhz * 1e6 * bytes_per_transfer * channels
bandwidth_gbs = bandwidth_bytes_per_second / (1<<30) # ~83.446 GiB/s
```

**Comment by hitorilabs** [comment]
Benchmarking imagenet + regular torch dataloader (unprocessed) + image folder vs. HF datataset
`batch_size=32, workers=16`

testing imagenet dataloading dataloader (unprocessed) + image folder vs. huggingface

*side note: scanning some random image and label samples in here, this dataset is pretty terrible by today's standards + why tf is there a single image at idx 1159338 that is RGBA and a couple grayscale images just to ruin my life*

**Comment by hitorilabs** [comment]
Testing out HDF5 as a data format for training data, but it was pretty annoying trying to figure out how to quickly write all of imagenet into the single file.

First attempt at fixing the transform I didn't realize there was an RGBA image just chilling inside of `imagenet-1k`, so raised an exception bc I was expecting 3 channels instead of 4. I couldn't find any good suggestions for this, so I just tried to slice the tensor.

```
tv_transforms = T.Compose([
    T.Resize(opt.imageSize),
    T.CenterCrop(opt.imageSize),
    T.ToImage(),
    T.ToDtype(torch.float32, scale=True),
    T.Lambda(lambda x: x[:3]),
    T.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5)),
    ])
```

Then it was raising errors because there were grayscale images with just 1 channel it basically getting an index error (implicit data not resizable error). I figured there had to be a better option so I looked a little harder and found that probably the cleanest solution is to do the conversion while it's still a PIL image.

```
tv_transforms = T.Compose([
    T.Lambda(lambda x: x.convert("RGB")),
    T.Resize(opt.imageSize),
    T.CenterCrop(opt.imageSize),
    T.ToImage(),
    T.ToDtype(torch.float32, scale=True),
    T.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5)),
    ])
```

In terms of making it go fast, the config I ended up running was `workers=16, batch_size=1024` (note, just processing would get `~5K img/s`) - this works out to about ~30GB of RAM usage. Large batches didn't work well presumably because of CPU memory bottlenecking - the actual file write itself is relatively trivial, so the synchronization isn't going to completely ruin performance on a regular 16 core machine (~500ms write vs. ~2s processing) so more frequent writes didn't end up being much of an issue. (total runtime is ~8.5 mins)

Screenshot 2023-12-31 at 12 28 06 PM

**Comment by hitorilabs** [comment]
Very naive benchmark, but HDF5 is pretty solid. I bet it's basically the performance of loading from a single file mmap'd numpy array, but with a more sane interface. The performance of just data access is pretty much the same w/ and w/o moving data onto GPU. I was surprised that by `BS=8192` it actually finished loading all of `imagenet-1k` before flushing the buffer for the first time (~100s total). 

There's plenty of bandwidth left to fill for transfers to GPU, so it's really just being limited by being single-threaded (i'm pretty sure parallel reads are straightforward). 

For reference, the original files for imagenet-1k come out to `~160GB` and the pre-processed HDF5 file is `~66GB`

*note: every image was processed into* `shape=(3, 64, 64), dtype=float32`

Screenshot 2023-12-31 at 3 23 38 PM

**Comment by hitorilabs** [comment]
`danbooru2021` dataset incoming

EDIT: it has been downloading since and I only got 500GB so far

**Comment by hitorilabs** [comment]
Clearly loading from disk is a **serious** bottleneck I always knew this, but I never really felt the gap. I literally have one of the fastest read speeds you're going to get over PCIE 4.0 x4 (WD Black SN850X @ 7300MB/s read) and it is still too slow.

You're looking at best 6K img/s reading off disk vs. upwards of 300K img/s off RAM (note: ~55s to load into RAM)

**Comment by hitorilabs** [comment]
After all this effort, I've found that np.memmap through torch dataloader just works pretty seamlessly w/ data loader. It is 4x faster than HDF5 (mainly due to enabling parallelism) without having to pre-load everything into RAM.

~100k img/s once it's "warmed up" + is way less harsh on the CPU (as indicated by htop + my CPU fans)

83708693-ed36-4926-b269-bd25ec92d5b9-2 jpg

many such cases

**Comment by hitorilabs** [comment]
The dataloader has to be literally the biggest mystery ever... just out of curiousity I wanted to check if the torch dataloader would get in the way of trying to get 300K img/s when I pre-load the dataset into RAM.

- I just tried pre-loading the memmap'd array into RAM w/ 8 workers (it was terrible 30K img/s at some point, but it doesn't really make sense to put the whole dataset in RAM and also let the dataloader buffer it into RAM again) 
- Then I suspected it's due to whatever multi-process initialization causing it to be slow, so I changed it to 1 worker (assuming this gets around the problem, but still terrible ~50K img/s)
- Then I changed it to 0 workers because I assume this might have a behaviour more similar to loading data in main process, and it yielded 200K img/s, but still not quite the 300K img/s that I know is possible with a regular loop in the main process. Additionally, smaller batch sizes seemed to let me get higher throughput, but practically speaking when writing an actual training loop - this will probably be a terrible idea.

**Comment by hitorilabs** [comment]
closing this out for now, will be actually doing the training loops and stuff like that - not stalling anymore

EDIT: I was naive it's also not exactly accurate (cuda sync and what not)
